### PR TITLE
Cw 1362/import

### PR DESCRIPTION
**Pull request by tejaswinichile**
Fixes: https://linear.app/chatwoot/issue/CW-1362/csv-imports-are-not-working-properly

### CODE REVIEW

**Comment by tejaswinichile** [comment]
Screenshot 2023-03-28 at 12 52 30 PM

Phone number import as an column value not as an custom attr

**Comment by pranavrajs** [comment]
@tejaswinichile @sojan-official Can we prioritize getting this out? There are multiple reports on the contact import in the support last week.

File: app/views/mailers/administrator_notifications/channel_notifications_mailer/erroneous_import_records.liquid [hunk 1]
```diff
@@ -0,0 +1,10 @@
+<p>Hello,</p>
+
+<p>Your contacts import upload failed, please check the attached file for the failed records. </p>
+<p>Please check the contacts tab.</p>
+
```
Comment (sojan-official) [inline app/views/mailers/administrator_notifications/channel_notifications_mailer/erroneous_import_records.liquid:5]: lets include the count of how many records were successfully imported and how many failed . ref: https://linear.app/chatwoot/issue/CW-1362#comment-97daed6c

Also lets go with the something like following copy 

```
Your contact import has been completed. Please check the contacts tab to view the imported contacts
Number of records imported: count
{% if any records filed %}
Number of record failed: count
attachments
{% end % }

```

File: spec/jobs/data_import_job_spec.rb [hunk 2]
```diff
@@ -13,9 +14,23 @@
 
   it 'imports data into the account' do
     csv_length = CSV.parse(data_import.import_file.download, headers: true).length
+
     described_class.perform_now(data_import)
     expect(data_import.account.contacts.count).to eq(csv_length)
     expect(data_import.reload.total_records).to eq(csv_length)
     expect(data_import.reload.processed_records).to eq(csv_length)
   end
+
```
Comment (sojan-official) [inline spec/jobs/data_import_job_spec.rb:23]: Let's add specific test cases that we should have included earlier. 

- If existing email contact exists the data gets updated
- if existing phone number contact exists data gets updates
-  imported contacts count / specific attributes are updated correctly

## edge cases
- csv contains a row with email 'xyc` and phone number `123` . the database has contact 1 with email `xyc` and contact 2 with phone_number `123`  .. which contact gets updated ?

**Review by sojan-official — COMMENTED** [review]
Added comments. 
TLDR;
- [ ] Let's perform a benchmark 
- [x] Let's update specs for cases we haven't covered  before
- [ ] Lets look at options to optimise the speed.

**Comment by sojan-official** [comment]
note: lets also support updating the  `company` attribute ref: https://github.com/chatwoot/chatwoot/issues/3462#issuecomment-991217378

File: app/jobs/data_import_job.rb [hunk 3]
```diff
@@ -6,11 +6,17 @@ class DataImportJob < ApplicationJob
...
-    result = Contact.import contacts, on_duplicate_key_update: :all, batch_size: 1000
-    data_import.update!(status: :completed, processed_records: csv.length - result.failed_instances.length, total_records: csv.length)
+    rejected_contacts = []
+    @data_import = data_import
+    @data_import.update!(status: :processing)
+    csv = CSV.parse(@data_import.import_file.download, headers: true)
+    csv.each { |row| contacts << build_contact(row.to_h.with_indifferent_access, @data_import.account) }
+    contacts.each_slice(1000) do |contact_chunks|
+      rejected_contacts << contact_chunks.reject { |contact| contact.valid? && contact.save! }
+    end
```
[the remark this answers is not in the document]
Comment (tejaswinichile) [reply app/jobs/data_import_job.rb:16]: @sojan-official Around 10000 records take 1-1.37 minutes.

File: app/jobs/data_import_job.rb [hunk 4]
```diff
@@ -6,11 +6,24 @@ class DataImportJob < ApplicationJob
...
+    @data_import.update!(status: :processing)
+    csv = CSV.parse(@data_import.import_file.download, headers: true)
+
+    csv.each do |row|
+      current_contact = build_contact(row.to_h.with_indifferent_access, @data_import.account)
+      current_contact.valid? ? contacts << current_contact : rejected_contacts << current_contact
+    end
+
+    # <struct ActiveRecord::Import::Result failed_instances=[], num_inserts=1, ids=[444, 445], results=[]>
+    Contact.import(contacts, synchronise: contacts, on_duplicate_key_ignore: true, track_validation_failures: true, validate: true,
```
Comment (sojan-official) [inline app/jobs/data_import_job.rb:20]: Note: I will take a pass at `failed_instances` and see if we can get it populated.

**Review by sojan-official — COMMENTED** [review]
code LGTM. will do the testing and get this merged.

File: app/jobs/data_import_job.rb [hunk 5]
```diff
@@ -5,53 +5,118 @@ class DataImportJob < ApplicationJob
...
+    Contact.import(contacts, synchronize: contacts, on_duplicate_key_ignore: true, track_validation_failures: true, validate: true, batch_size: 1000)
+  end
+
+  def update_data_import_status(processed_records, rejected_records)
+    @data_import.update!(status: :completed, processed_records: processed_records, total_records: processed_records + rejected_records)
+  end
+
+  def build_contact(params, account)
+    contact = find_or_initialize_contact(params, account)
     contact.name = params[:name] if params[:name].present?
```
Comment (sojan-official) [inline app/jobs/data_import_job.rb:49]: Let's assign `company` , and `city` as well. Here in additional attributes. 
let's also look at whether we can add country as well.

File: app/jobs/data_import_job.rb [hunk 6]
```diff
@@ -5,53 +5,118 @@ class DataImportJob < ApplicationJob
...
 
+  def parse_csv_and_build_contacts
+    contacts = []
+    rejected_contacts = []
+    csv = CSV.parse(@data_import.import_file.download, headers: true)
+
+    csv.each do |row|
+      current_contact = build_contact(row.to_h.with_indifferent_access, @data_import.account)
+      current_contact.valid? ? contacts << current_contact : rejected_contacts << current_contact
+    end
```
Comment (sojan-official) [inline app/jobs/data_import_job.rb:33]: rejected_contacts << row + current_contact.errors

File: app/jobs/data_import_job.rb [hunk 7]
```diff
@@ -5,53 +5,118 @@ class DataImportJob < ApplicationJob
...
+    return unless rejected_contacts.any?
+
+    csv_data = generate_csv_data(rejected_contacts)
+    @data_import.failed_records.attach(io: StringIO.new(csv_data), filename: "#{Time.zone.today.strftime('%Y%m%d')}_contacts.csv",
+                                       content_type: 'text/csv')
+  end
+
+  def generate_csv_data(rejected_contacts)
+    CSV.generate do |csv|
+      csv << %w[id first_name last_name email phone_number identifier gender errors]
```
Comment (sojan-official) [inline app/jobs/data_import_job.rb:103]: original csv headers + errors

**Review by sojan-official — COMMENTED** [review]
minor comments
